File: cdm/java/mil/tatrc/physiology/datamodel/system/physiology/SEPupillaryResponse.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package mil.tatrc.physiology.datamodel.system.physiology;

import com.kitware.physiology.cdm.Physiology.PupillaryResponseData;

import mil.tatrc.physiology.datamodel.properties.SEScalarNegative1To1;

public class SEPupillaryResponse
{
  protected SEScalarNegative1To1  reactivityModifier;
  protected SEScalarNegative1To1  shapeModifier;
  protected SEScalarNegative1To1  sizeModifier;
  
  public SEPupillaryResponse()
  {
    reactivityModifier = null;
    shapeModifier = null;
    sizeModifier = null;
  }

  public void reset()
  {
    if (reactivityModifier != null)
      reactivityModifier.invalidate();
    if (shapeModifier != null)
      shapeModifier.invalidate();
    if (sizeModifier != null)
      sizeModifier.invalidate();
  }

  public static void load(PupillaryResponseData src, SEPupillaryResponse dst)
  {
    dst.reset();
    if (src.hasReactivityModifier())
      SEScalarNegative1To1.load(src.getReactivityModifier(),dst.getReactivityModifier());
    if (src.hasShapeModifier())
      SEScalarNegative1To1.load(src.getShapeModifier(),dst.getShapeModifier());
    if (src.hasSizeModifier())
      SEScalarNegative1To1.load(src.getSizeModifier(),dst.getSizeModifier());    
  }

  public static PupillaryResponseData unload(SEPupillaryResponse src)
  {
    PupillaryResponseData.Builder dst = PupillaryResponseData.newBuilder();
    SEPupillaryResponse.unload(src,dst);
    return dst.build();
  }

  protected static void unload(SEPupillaryResponse src, PupillaryResponseData.Builder dst)
  {
    if (src.reactivityModifier != null)
      dst.setReactivityModifier(SEScalarNegative1To1.unload(src.reactivityModifier));
    if (src.shapeModifier != null)
      dst.setShapeModifier(SEScalarNegative1To1.unload(src.shapeModifier));
    if (src.sizeModifier != null)
      dst.setSizeModifier(SEScalarNegative1To1.unload(src.sizeModifier));    
  }
  
  public SEScalarNegative1To1 getReactivityModifier()
  {
    if (reactivityModifier == null)
      reactivityModifier = new SEScalarNegative1To1();
    return reactivityModifier;
  }
  public boolean hasReactivityModifier()
  {
    return reactivityModifier == null ? false : reactivityModifier.isValid();
  }
  
  public SEScalarNegative1To1 getShapeModifier()
  {
    if (shapeModifier == null)
      shapeModifier = new SEScalarNegative1To1();
    return shapeModifier;
  }
  public boolean hasShapeModifier()
  {
    return shapeModifier == null ? false : shapeModifier.isValid();
  }
  
  public SEScalarNegative1To1 getSizeModifier()
  {
    if (sizeModifier == null)
      sizeModifier = new SEScalarNegative1To1();
    return sizeModifier;
  }
  public boolean hasSizeModifier()
  {
    return sizeModifier == null ? false : sizeModifier.isValid();
  }
}

File: cdm/cpp/system/physiology/SEPupillaryResponse.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/

#include "stdafx.h"
#include "system/physiology/SEPupillaryResponse.h"
#include "properties/SEScalarNegative1To1.h"

SEPupillaryResponse::SEPupillaryResponse(Logger* logger)
{
  m_ReactivityModifier = nullptr;
  m_ShapeModifier = nullptr;
  m_SizeModifier = nullptr;
}

SEPupillaryResponse::~SEPupillaryResponse()
{
  Clear();
}

void SEPupillaryResponse::Clear()
{
  SAFE_DELETE(m_ReactivityModifier);
  SAFE_DELETE(m_ShapeModifier);
  SAFE_DELETE(m_SizeModifier);
}

const SEScalar* SEPupillaryResponse::GetScalar(const std::string& name)
{
  if (name.compare("ReactivityModifier") == 0)
    return &GetReactivityModifier();
  if (name.compare("ShapeModifier") == 0)
    return &GetShapeModifier();
  if (name.compare("SizeModifier") == 0)
    return &GetSizeModifier();
  return nullptr;
}

void SEPupillaryResponse::Load(const cdm::PupillaryResponseData& src, SEPupillaryResponse& dst)
{
  SEPupillaryResponse::Serialize(src, dst);
}
void SEPupillaryResponse::Serialize(const cdm::PupillaryResponseData& src, SEPupillaryResponse& dst)
{
  dst.Clear();
  if (src.has_reactivitymodifier())
    SEScalarNegative1To1::Load(src.reactivitymodifier(), dst.GetReactivityModifier());
  if (src.has_shapemodifier())
    SEScalarNegative1To1::Load(src.shapemodifier(), dst.GetShapeModifier());
  if (src.has_sizemodifier())
    SEScalarNegative1To1::Load(src.sizemodifier(), dst.GetSizeModifier());
}

cdm::PupillaryResponseData* SEPupillaryResponse::Unload(const SEPupillaryResponse& src)
{
  cdm::PupillaryResponseData* dst = new cdm::PupillaryResponseData();
  SEPupillaryResponse::Serialize(src, *dst);
  return dst;
}
void SEPupillaryResponse::Serialize(const SEPupillaryResponse& src, cdm::PupillaryResponseData& dst)
{
  if (src.HasReactivityModifier())
    dst.set_allocated_reactivitymodifier(SEScalarNegative1To1::Unload(*src.m_ReactivityModifier));
  if (src.HasShapeModifier())
    dst.set_allocated_shapemodifier(SEScalarNegative1To1::Unload(*src.m_ShapeModifier));
  if (src.HasSizeModifier())
    dst.set_allocated_sizemodifier(SEScalarNegative1To1::Unload(*src.m_SizeModifier));
}

bool SEPupillaryResponse::HasReactivityModifier() const
{
  return m_ReactivityModifier == nullptr ? false : m_ReactivityModifier->IsValid();
}
SEScalarNegative1To1& SEPupillaryResponse::GetReactivityModifier()
{
  if (m_ReactivityModifier == nullptr)
    m_ReactivityModifier = new SEScalarNegative1To1();
  return *m_ReactivityModifier;
}
double SEPupillaryResponse::GetReactivityModifier() const
{
  if (m_ReactivityModifier == nullptr)
    return SEScalar::dNaN();
  return m_ReactivityModifier->GetValue();
}

bool SEPupillaryResponse::HasShapeModifier() const
{
  return m_ShapeModifier == nullptr ? false : m_ShapeModifier->IsValid();
}
SEScalarNegative1To1& SEPupillaryResponse::GetShapeModifier()
{
  if (m_ShapeModifier == nullptr)
    m_ShapeModifier = new SEScalarNegative1To1();
  return *m_ShapeModifier;
}
double SEPupillaryResponse::GetShapeModifier() const
{
  if (m_ShapeModifier == nullptr)
    return SEScalar::dNaN();
  return m_ShapeModifier->GetValue();
}

bool SEPupillaryResponse::HasSizeModifier() const
{
  return m_SizeModifier == nullptr ? false : m_SizeModifier->IsValid();
}
SEScalarNegative1To1& SEPupillaryResponse::GetSizeModifier()
{
  if (m_SizeModifier == nullptr)
    m_SizeModifier = new SEScalarNegative1To1();
  return *m_SizeModifier;
}
double SEPupillaryResponse::GetSizeModifier() const
{
  if (m_SizeModifier == nullptr)
    return SEScalar::dNaN();
  return m_SizeModifier->GetValue();
}


